namer/namer.cc

```diff
@@ -1191,26 +1192,30 @@ class SymbolDefiner {
                 behaviorLocs.emplace_back(ctx.locAt(klass.declLoc));
                 symbol.data(ctx)->flags.isBehaviorDefining = true;
             }
-        }
-        auto singletonClass = symbol.data(ctx)->singletonClass(ctx); // force singleton class into existence
-        if (symbol != core::Symbols::root() && !isUnknown) {
-            singletonClass.data(ctx)->addLoc(ctx, ctx.locAt(klass.declLoc));
-        }
 
-        // This willDeleteOldDefs condition is a hack to improve performance when editing within a method body.
-        // Ideally, we would be able to make finalizeSymbols fast/incremental enough to run on all edits.
-        if (willDeleteOldDefs) {
-            // Reset resultType to nullptr for idempotency on the fast path--it will always be
-            // re-entered in resolver.
-            symbol.data(ctx)->resultType = nullptr;
-            singletonClass.data(ctx)->resultType = nullptr;
-            // TODO(jez) This is a gross hack. We are basically re-implementing the logic in
-            // singletonClass to reset the <AttachedClass> type template to what it used to be.
-            // Is there a better way to accomplish this? (This is largely the same as the bad locs problem above;
-            // we can probably be more principled about what state calling `singletonClass` sets up/resets.)
-            auto todo = core::make_type<core::ClassType>(core::Symbols::todo());
-            auto tp = singletonClass.data(ctx)->members()[core::Names::Constants::AttachedClass()].asTypeMemberRef();
-            tp.data(ctx)->resultType = core::make_type<core::LambdaParam>(tp, todo, todo);
```

Comment:
Why was all this moved inside the `if (symbol != core::Symbols::root() && !isUnknown)`? I understand the addition of the `if (!isModule) {` check, but I don't understand this move.

Reply:
Yeah it's a bit wacky, sorry.

We can go piece by piece. Referring to the old line numbers:

- 1195 needs to go into the condition because the PR is changing it so that `singletonClass` is only ever called once we know `!isUnknown`.

  We also know that calling `singletonClass` on `root` would have been a no-op anyways, because that singleton class is always created by `initEmpty`. So for root, it doesn't matter, and for non-root, we need to guard it by `!isUnknown`. So the `singletonClass` call can move in.

- On lines 1196 to 1198, well, that's the same condition, so just move it in there.

- Then to understand the `willDeleteOldDefs` part, we have to understand what it's really trying to do.

  I added this code as an absolute hack ("This is a gross hack"). Basically, there is some **other** state that calling `singletonClass` is supposed to initialize for the first time, specifically around resultTypes and type members. Later, resolver sees that things like resultTypes and type members have not been resolved yet, and does work to correctly initialize those fields.
  
  So all this `willDeleteOldDefs` logic is attempting to re-implement half of the call to `singletonClass` here in namer. That's what makes it gross. You can look at the implementation of `ClassOrModule::singletonClass` to see the similarities.

  But what it means is that we only have to run this logic on the fast path whenever `singletonClass` would have been called on the slow path. And since that call to `singletonClass` from 1195 has moved into the condition, well now we can (have to) also move all this code into the condition.

A future, non-gross hack would maybe be to make calling `singletonClass` reset this information unconditionally, or to split `singletonClass` into two methods, and audit all calls to `singletonClass` to see if they need to be rewritten to call `singletonClass` + the new helper. etc. But I was lazy and just wrote this hack, unfortunately.